### ISSUE TITLE
prisma generate, typings not updating in VSCode unless IDE is restarted.

### ISSUE THREAD

**Issue by dstampher** [prisma/prisma#14722]
### Bug description

Changing the type of a field in a model, or adding a new field, then running the command `npx prisma db push` and `npx prisma generate` are not updating the typings in my IDE, unless I restart VSCode after making my schema changes and running those commands.  

During prototyping where I am frequently updating my models, this disrupts my workflow significantly.

### How to reproduce

1. Add a new field to a model
2. Run `npx prisma db push` and `npx prisma generate`
3. The field created in step 1 will not be recognized when executing await prisma.model.create(data: {...}).
4. Restart VSCode
5. Step 3 error goes away.

### Expected behavior

The IDE recognizes the updates to the model and does not require a restart.

### Prisma information

```
<!-- Do not include your database credentials when sharing your Prisma schema! -->
// This is your Prisma schema file,
// learn more about it in the docs: https://pris.ly/d/prisma-schema

generator client {
  provider = "prisma-client-js"
}

datasource db {
  provider = "postgresql"
  url      = env("DATABASE_URL")
}

model YoutubeAccount {
  username        String          @unique
  password        String
  profile_path    String
  proxy           String
  video_templates VideoTemplate[]
}

model VideoTemplate {
  id                     Int             @id @default(autoincrement())
  inputDirs              String[]
  title                  String
  description            String
  tags                   String[]
  clipCount              Int
  duration               Int
  width                  Int
  height                 Int
  YoutubeAccount         YoutubeAccount? @relation(fields: [youtubeAccountUsername], references: [username])
  youtubeAccountUsername String?
}
```

### Environment & setup

- OS: 
- Windows
- Database: 
- PostgreSQL
- Node.js version: 
- Node 18.4.0

### Prisma Version

```
npx prisma -v
Environment variables loaded from .env
prisma                  : 4.2.0
@prisma/client          : 4.2.0
Current platform        : windows
Query Engine (Node-API) : libquery-engine 2920a97877e12e055c1333079b8d19cee7f33826 (at node_modules\@prisma\engines\query_engine-windows.dll.node)
Migration Engine        : migration-engine-cli 2920a97877e12e055c1333079b8d19cee7f33826 (at node_modules\@prisma\engines\migration-engine-windows.exe)
Introspection Engine    : introspection-core 2920a97877e12e055c1333079b8d19cee7f33826 (at node_modules\@prisma\engines\introspection-engine-windows.exe)
Format Binary           : prisma-fmt 2920a97877e12e055c1333079b8d19cee7f33826 (at 
node_modules\@prisma\engines\prisma-fmt-windows.exe)
Default Engines Hash    : 2920a97877e12e055c1333079b8d19cee7f33826
Studio                  : 0.469.0
```

**Comment by ZiadJ** [comment]
Opening up the definition file by pressing F12 over a referenced type will fix it so there's not need to restart. Not the best solution I know but I thought I'd mention it until a real fix comes up.

**Comment by dstampher** [comment]
@ZiadJ Thanks for the pro-tip, that will make this 50% less annoying for now.  But 100% would be great.

**Comment by janpio** [comment]
Do you have the Prisma VSCode extension installed? We have some code that tries to work around this problem (which btw is not directly Prisma related, but just how the TS server works unfortunately) by restarting the TS server when we detect a change of the types.

**Comment by Jolg42** [comment]
Note, like Jan said, the https://marketplace.visualstudio.com/items?itemName=Prisma.prisma handles that automatically.

Also, you don't need to restart VS Code, this one from the command palette will do it perfectly: "TypeScript: Restart TS Server"
See https://stackoverflow.com/a/64456017/1345244 for example and shortcut.

**Comment by dstampher** [comment]
> 

Hi, just wanted to let you know that I do have the Prisma extension added to VS Code.  @Jolg42 Good to know about the steps you mentioned to also fix this, thanks.

**Comment by janpio** [comment]
Interesting, that means that this feature does not necessarily work as we hoped it would 😢

**Comment by dstampher** [comment]
@janpio Let me know if there is any other info I can provide that would be of assistance.  I assume you are not having this problem on your end, when the extension is installed.  

For now I found some decent workarounds thanks to other commenters here. :)

**Comment by janpio** [comment]
I personally can actually reproduce this sometimes, but have no idea how it happens and other colleagues can not. We are pretty clueless right now.

**Comment by younes-alouani** [comment]
This bug is easy to reproduce. 
Just add a new field to any model that has been already pushed to database. Push the new model to database. 
Now the VS CODE will show an error when you do any operation using PRISMA with the new field. It's says that the field doesn't belong to Prisma.

**Comment by joaopedrodcf** [comment]
The same issue happening to me in a side project, the solution as said before was every time we change the `schema.prisma` in vscode  we jus run the task: `Developer: Reload window`.

We use in our project this script to run in dev nextjs and prisma watcher:

```
"dev": "concurrently \"prisma generate --watch\" \"next dev\"",
```

Example removing this location property from the `schema.prisma`
<img width="560" alt="image" src="https://user-images.githubusercontent.com/11476152/206048328-e817927d-257c-4041-a2bb-eb0d8d9a8b84.png">


Will show correctly in our terminal
<img width="726" alt="image" src="https://user-images.githubusercontent.com/11476152/206048193-1b6e172c-549a-4d7c-a269-1b0bcc3758d2.png">

But then in the code we can see there is no errors
<img width="416" alt="image" src="https://user-images.githubusercontent.com/11476152/206048563-3b51ed7f-260c-4efd-bcef-4bf05b2e1f29.png">

But by running the reload
<img width="613" alt="image" src="https://user-images.githubusercontent.com/11476152/206048626-ccf708b6-a447-4029-9c4a-b569eb391b88.png">

The error now appears
<img width="748" alt="image" src="https://user-images.githubusercontent.com/11476152/206048759-2cc93259-7a37-4d84-8f2f-d8acff183205.png">

If there was a way at least to automate vscode to everytime we change the `schema.prisma` to reload the window or at least restart the ts server would improve so much the deloveper experience 🙏

**Comment by Jolg42** [comment]
@joaopedrodcf If you install the [Prisma VS Code extension](https://marketplace.visualstudio.com/items?itemName=Prisma.prisma) it will automatically trigger the "Restart TS server" command when a change is detected.

@joaopedrodcf @younes-alouani Did you try Prisma's VS Code extension? Is it working for you?

Alternatively, you can also trigger the same command manually via the command palette:
<img width="523" alt="Screenshot 2022-12-07 at 10 01 55" src="https://user-images.githubusercontent.com/1328733/206135036-260c330d-7339-4070-82bf-c908d2e10fc8.png">

**Comment by DPangerl** [comment]
So I have installed The Prisma VS Code extension (v4.11.0). Though I have to restart the TS server manually for every change. Is there anything I have to activate for the extension to update automatically? 
Besides a checkbox to (de)activate `fileWatcher` (which is checked), there is only a trace:server setting

**Comment by janpio** [comment]
(That checkbox is terrible UX - I have no idea if it is supposed to work when checked, or unchecked. I think actually when unchecked. Please try that. (I raised this internally so we can improve the wording there))

**Comment by DPangerl** [comment]
@janpio Thank you for the hint…
I've just assumed, that the checkbox will activate watch mode. But I took a closer look today… It is indeed confusing :-D

So i tried it with the checkbox activated as well as deactivated, restarted VS Code before making changes to the schema file. But without luck. Maybe I'm doing it all wrong, so here is my workflow:

- Adding `NewModel` to schema.prisma
- `npx prisma db push`
(console says client is generated, all in sync)
- Adding a `prisma.newModel.findUnique()` call to a `ts`-file
- getting a typescript error because `newModel` does not exist
- restarting TS Server manually
- TS Error is gone

I won't go crazy with restarting the server manually all the time, but if the extension is expected to handle this, it shoud work ;-)

**Comment by janpio** [comment]
Yes, that is exactly the use case for that file watcher. Are you generating into a custom `output` folder maybe? That is a scenario it can not handle. Or some weird monorepo setup? Maybe create an issue over at https://github.com/prisma/language-tools and provide some more information that might help us debug this. Thanks.

**Comment by dvicagolf** [comment]
Idk if it's just me, but this sounds super straigtforward... is there an API inside of VS code to restart TS server manually? 

Can't we just chain that command to the generate one, since that's the one that's ran when the schema file changes are to be officially built?

**Comment by janpio** [comment]
That is exactly what our [VS Code extension](https://marketplace.visualstudio.com/items?itemName=Prisma.prisma) does - it reacts to the types being (re)generated via `prisma generate`, and then reloads the TS server so it definitely picks up the changes (manually you can do so via `TypeScript: Restart TS Server` in the comman palette). 

Unfortuantely, there is no way to trigger that reload without having our extension (so just via Prisma CLI) installed. If you see an option for that, happy to learn about it - would solve quite a bit of pain for all our users.

**Comment by dvicagolf** [comment]
I have the extension and didn't notice that it's doing so. Cool to know though, I'll be looking to see if I can reproduce it not being reloaded.

Doing it manually is a really small trade-off for how amazing this package is!

**Comment by janpio** [comment]
(We actually rewrote that logic recently, so it will probably only work properly with 4.15.0. [Before it was watching the type files that change implicitly when you `prisma generate` and was kinda broken on some operating systems, now it watches a file that the CLI writes when you `prisma generate` _explicitly_])

**Comment by hanayashiki** [comment]
Why is restarting typescript language server indeed needed? Watching file changes should work out of the box.

**Comment by janpio** [comment]
Tell that to VS Code and the TS language server :D 

We assume the problem is caused by Prisma generating the files to `node_modules/.prisma` by default, but have not figured out or confirmed for that to be the case. Let us know if using a custom `output` in the `generator` block makes the problem go away.

**Comment by hanayashiki** [comment]
> Tell that to VS Code and the TS language server. We certainly agree :D
> 
> We assume the problem is caused by Prisma generating the files to `node_modules/.prisma` by default, but have not figured out or confirmed for that to be the case. Let us know if using a custom `output` in the `generator` block makes the problem go away.

If you run "yarn build" and update your dependencies in the node_modules there's a good chance TS server will catch the changes. What is the difference when it comes to node_modules/.prisma?

**Comment by olivierlacan** [comment]
Sadly we've run into this issue repeatedly as well and had to internalize the need to run `prisma generate` whenever we start getting strange type errors after altering the schema, for example when we rename some models or fields. 

It is very likely due to the fact we use Yarn workspaces and our `node_modules` directory is actually located at the top level of a structure like this: 

```
apps/api/prisma/prisma.schema
node_modules/prisma/build
```

There's no obvious error from the Prisma LSP however that might indicate it's failing to find the relevant generated files in `node_modules/`. Just to be clear, a simple `prisma generate` in the CLI solves the issue immediately. 

Importantly `Restart TS Server` does *not* fix this issue for us under the latest VSCode on macOS 13.6 (22G120). 

https://github.com/prisma/prisma/assets/65950/f0f16907-9412-4cfc-ac56-1694d7aa6f52

**Comment by umerthow** [comment]
Same with me..

Try to uninstall and install the extention but still got issue. Wanna to restart TS but command pallet not appear .  I must be **manually delete my node_modules** file and install again. 

Very exhausted if every change prisma schema should manually delete node module.  

I'm use MAC OS Somona and latest VScode btw.

any other solution  work for this?

**Comment by Jolg42** [comment]
@olivierlacan This behavior is expected, I don't think it fits in this issue. You could create a new issue or discussion to discuss your developer experience, it would be interesting, I think.
Since the Prisma Client relies on code generation, after any change to the Prisma schema, you must call `prisma generate` to update it.
You could try `npx prisma generate --watch` in the background to automatically update the Prisma Client when the Prisma schema changes.
See docs: https://www.prisma.io/docs/reference/api-reference/command-reference#generate

**Comment by madipta** [comment]
it's happen to me after upgrade to prisma 5.9.0

**Comment by Jolg42** [comment]
Hi @madipta I think you are experiencing a regression with identified in `5.9.0`, could you check 
See https://github.com/prisma/prisma/issues/22870
You can try the following versions `5.10.0-integration-fix-nodenext.2` and `5.9.1-dev.1`
Both should fix the issue.

If you can let us know if that indeed fixes the problem for you, that would be great.
Stay tuned later for the official patch release.

**Comment by madipta** [comment]
Hi @Jolg42 I have tried 5.9.1-dev.1 and it works

**Comment by theHasanas** [comment]
I just run "Reload window" command on VSC; it's much faster than restarting or going to the file to reload.

For me it goes in this sequence:
1. Open command palette `Shift + Command + P` (Mac) `Ctrl + Shift + P` (Windows/Linux)
2. Type "Reload" (since I use it often, it's the first choice I get as soon as I type the letter `r`.
3. Click `Return`.

I am back into the regular flow using this in less than 3 seconds.

**Comment by Zakisb** [comment]
It's annoying to reload the window everytime. it slows the dev process. Why not provide a solution that detects changes and update automatically ?

**Comment by janpio** [comment]
See this comment from 2022 the mentions that we have that: https://github.com/prisma/prisma/issues/14722#issuecomment-1210301016

**Comment by chrisblakely01** [comment]
I have the VSC plugin installed and still hit this error. What works for me is deleting the `.prisma` folder in node_modules, then generating prisma client:

```
  "prisma:generate": "rimraf ./node_modules/.prisma && npx prisma generate"
```

not ideal as you (may) need rimraf. Would be nice if this was baked into the prisma command

```
npx prisma generate --clean
```

or something.

**Comment by tkeith** [comment]
Adding my data point in case it's helpful:

The extension is not solving this for me, and neither is `TypeScript: Restart TS server` or `TypeScript: Reload project`. The only thing that's resolving this for me is `Developer: Reload window`. Maybe this explains why the extension isn't resolving this if it's triggering the `TypeScript: Restart TS server` command. I do recall that `TypeScript: Restart TS server` used to resolve the issue for me as well, but it might be a change (in VSCode or the TypeScript integration) in the last year or so that broke this.

Prisma extension version: v5.12.1

IDE info (Cursor):

```
Version: 0.31.1
VSCode Version: 1.86.2
Commit: e5a039914ce9ebe93667ab7248a02d546f9bb760
Date: 2024-04-02T06:08:50.662Z (1 wk ago)
Electron: 27.2.3
ElectronBuildId: undefined
Chromium: 118.0.5993.159
Node.js: 18.17.1
V8: 11.8.172.18-electron.0
OS: Darwin arm64 23.4.0
```

**Comment by abhishekbhardwaj** [comment]
Just as @tkeith said. Encountering the exact same issue and only `Developer: Reload window` seems to be working.

To be honest, this seems to have broken with the latest Prisma or VSCode update. not sure which tbh. It was working perfectly up until a few days ago.

**Comment by abhishekbhardwaj** [comment]
@Jolg42 Yes - that's what I usually do. Kinda realized something else - @tkeith - this seems to be a Cursor problem.
I switched to Code (latest version) for a day and `Typescript: Restart TS Server` seems to be doing the trick everytime. But doesn't seem to do anything on Cursor.

**Comment by c5n8** [comment]
In my case, switching devcontainer image from alpine to debian fixed it.

**Comment by c5n8** [comment]
In my case, using custom output path fixed it.
https://www.prisma.io/docs/orm/prisma-client/setup-and-configuration/generating-prisma-client#using-a-custom-output-path

**Comment by jason-ezenwa** [comment]
> @joaopedrodcf If you install the [Prisma VS Code extension](https://marketplace.visualstudio.com/items?itemName=Prisma.prisma) it will automatically trigger the "Restart TS server" command when a change is detected.
> 
> @joaopedrodcf @younes-alouani Did you try Prisma's VS Code extension? Is it working for you?
> 
> Alternatively, you can also trigger the same command manually via the command palette: <img alt="Screenshot 2022-12-07 at 10 01 55" width="523" src="https://user-images.githubusercontent.com/1328733/206135036-260c330d-7339-4070-82bf-c908d2e10fc8.png">

just wanted to note that downloading the extension does not solve the issue for me, but manually restarting ts server does the trick.

**Comment by davidbcarr** [comment]
I was experiencing this issue, but none of the proposed workaround/fixes were helping.

Upon further investigation it was caused by something else - **I was extending my types in other parts of my code**, which was causing the intellisense to give me the properties from my type extension, rather than from the Prisma schema.

Sharing here in case anyone else is in the same (silly) boat as me!

**Comment by theonlydaleking** [comment]
For anyone else that got this far - one thing that was causing an issue for me was the `VSCode typescript nightly` extension. prisma generate would crash the TS server and it wouldn't restart. 

Uninstalling this extension - just using the default TS works now.

**Comment by aaronvg** [comment]
hey everyone, we forked Prisma a while back and made our own language: [BAML](https://github.com/BoundaryML/baml) -- and we managed to fix this issue, though not sure if it'll work for Prisma.

At least for our implementation of the `baml_client` (similar to `prisma_client`), we generate a directory atomically using a `rename` operation (unsure if Prisma uses this technique). Most filewatchers are really bad at detecting renames, so we had to nudge the modifiedTime of the files for TS Server to pick up the changes. This allowed us to stop restarting the TS server on every generation.

Here is the PR: https://github.com/BoundaryML/baml/pull/1183/files 

BTW, would love to chat at some point @janpio if you ever have the time!
